### PR TITLE
Allow Frustum to transition between Equiangular Maps

### PR DESCRIPTION
**Pull request by mar-celine**
## Proposed changes

Currently Equiangular Mapping is disabled in the BinaryCompactObject domain.
The reason is because a transition mapping is needed in order to go from a full equiangular map in the Frustums around a single excision surface, to a half equiangular map in the half Wedges that make out the outer spherical layers of blocks.

The map in this PR uses the Frustum map to interpolate between a half equiangular
map at the upper face to a full equiangular map at the lower face. The bulging-out capability of the Frustum map is preserved - one can both bulge out the map and use this half-to-full ability at the same time.

This partially addresses point ii of #3545 , a future PR will update the BinaryCompactObject domain to make use of this addition.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Domain/CoordinateMaps/Frustum.hpp [hunk 1]
```diff
@@ -226,6 +226,35 @@ namespace CoordinateMaps {
  * 0 & 0 & 0 \\
  * \end{bmatrix}.
  * \f]
+ *
+ * ### Using the Frustum Transition between Equiangular Maps
+ * Using the parameter \f$\phi\f$, the user can specify whether to use the
+ * full equiangular map on the upper +z face, or whether to use only the upper
+ * or lower half of the equiangular map on the upper +z face. This capability
+ * is used in the BinaryCompactObjectDomain, where two equiangular maps around
```
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.hpp]: the class name is just `BinaryCompactObject` (fix so Doxygen finds the link)

File: src/Domain/CoordinateMaps/Frustum.hpp [hunk 2]
```diff
@@ -226,6 +226,35 @@ namespace CoordinateMaps {
...
+ * each spherical object must be transitioned into a single equiangular map at
+ * the outer boundary. The transition region that interpolates between these
+ * two regions is the layer of Blocks with the Frustum maps.
+ *
+ * The full map is obtained by replacing setting
```
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.hpp]: replacing or setting?

File: src/Domain/CoordinateMaps/Frustum.hpp [hunk 3]
```diff
@@ -226,6 +226,35 @@ namespace CoordinateMaps {
...
+ * \f$\Xi = \frac{1 + \mathrm{Z}}{2}\Xi_{\mathrm{upper}}(\xi,\zeta) +
+ *          \frac{1 - \mathrm{Z}}{2}\Xi_{0}(\xi)\f$
+ *
+ * for the volume map \f$\vec{x}(\xi,\eta,\zeta)\f$, and
+ * \f$\Xi = \Xi_{\mathrm{upper}}\f$ for the surface map
+ * \f$\vec{\sigma}(\xi,\eta)\f$, where \f$\Xi_{\mathrm{upper}}\f$ is the
+ * \f$\phi\f$-dependent generalized equiangular map, which corresponds to the
+ * lower half of the equiangular map when equal to -1, and to the upper half of
```
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.hpp]: ... when _phi is_ equal to ... ?

File: tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp [hunk 4]
```diff
@@ -58,7 +58,7 @@ void test_suite_for_frustum(const bool with_equiangular_map) {
          {{upper_x_upper_base, upper_y_upper_base}}}};
     const CoordinateMaps::Frustum frustum_map(face_vertices, -1.0, 2.0, map_i(),
                                               with_equiangular_map, 1.2, false,
-                                              0.5);
+                                              1.0, 1.0);
```
Comment (nilsvu) [inline tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp:67]: 1. Is it possible to add a few more tests, e.g., to make sure the map is actually equiangular, that the bulging still works, etc?
2. Two of the CI runs failed the root finding.

**Review by nilsvu — CHANGES_REQUESTED** [review]
Looks pretty equiangular to me, nice! I'll run some convergence tests with this next 👍

Bildschirmfoto 2022-03-24 um 14 48 19

File: tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp [hunk 5]
```diff
@@ -56,7 +56,7 @@ void test_suite_for_frustum(const bool with_equiangular_map) {
          {{upper_x_lower_base, upper_y_lower_base}},
          {{lower_x_upper_base, lower_y_upper_base}},
          {{upper_x_upper_base, upper_y_upper_base}}}};
-    const CoordinateMaps::Frustum frustum_map(face_vertices, -1.0, 2.0, map_i(),
```
Comment (nilsvu) [inline tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp:59]: Why did you change this?

File: tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp [hunk 6]
```diff
@@ -302,6 +302,196 @@ void test_bulged_frustum_inv_map() {
...
+  const std::array<double, 3> zeroxizeroeta{{0.0, 0.0, 1.0}};
+  const std::array<double, 3> upperxizeroeta{{1.0, 0.0, 1.0}};
+
+  const std::array<double, 3> lowerxiuppereta{{-1.0, 1.0, 1.0}};
+  const std::array<double, 3> zeroxiuppereta{{0.0, 1.0, 1.0}};
+  const std::array<double, 3> upperxiuppereta{{1.0, 1.0, 1.0}};
+
+  // If Frustum is successfully bulged, points on upper +zeta face
+  // should have the same radius:
+  CHECK(magnitude(map(lowerxilowereta)) == approx(4.0 * sqrt(3.0)));
```
Comment (nilsvu) [inline tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp]: Perhaps put the points defined above in a std::vector so you can easily check them all?

Also, for readability it would be useful to define `radius = 4 * sqrt(3)` and also use it in the tests below.

File: tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp [hunk 7]
```diff
@@ -302,6 +302,196 @@ void test_bulged_frustum_inv_map() {
...
+       map.jacobian(lowerxizeroeta).get(2, 0)}};
+  const std::array<double, 3> zeroxizeroeta_dx_dxi{
+      {map.jacobian(zeroxizeroeta).get(0, 0),
+       map.jacobian(zeroxizeroeta).get(1, 0),
+       map.jacobian(zeroxizeroeta).get(2, 0)}};
+  const std::array<double, 3> upperxizeroeta_dx_dxi{
+      {map.jacobian(upperxizeroeta).get(0, 0),
+       map.jacobian(upperxizeroeta).get(1, 0),
+       map.jacobian(upperxizeroeta).get(2, 0)}};
+  CHECK(magnitude(lowerxizeroeta_dx_dxi) == approx(4.0 * M_PI_4 * sqrt(3.0)));
```
Comment (nilsvu) [inline tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp]: Could you define a variable for this value please, and comment or make clearer (e.g. by using a `radius` variable) why this is the expected value? Same for the tests below.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp (hunk 5 above)
Comment (mar-celine) [reply tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp:59]: My hunch is that the original frustums had upper and lower bases that were very large relative to its height, causing the map shape to have very acute/obtuse angles and probably contributing to the bulged map numerical inverse failing for extreme cases. Making the frustum a little taller improves this.

The ratios of the lengths/bases/heights tested still cover a broad range, and the frustums used in the binarycompactobject domain fall comfortably in this range so I feel it's fine to restrict the range of frustums that are tested.

I can add a comment saying that "empirically, bulged frustums should have a ratio of base to height that is no greater than 6:1, at higher ratios the root finding begins to fail" or something like that

File: tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp [hunk 8]
```diff
@@ -302,6 +308,165 @@ void test_bulged_frustum_inv_map() {
...
+      {{{0.0, -1.0, 1.0}}, {{0.0, 0.0, 1.0}}, {{0.0, 1.0, 1.0}}}};
+
+  const std::array<std::array<double, 3>, 3> zeroetapts{
+      {{{-1.0, 0.0, 1.0}}, {{0.0, 0.0, 1.0}}, {{1.0, 0.0, 1.0}}}};
+
+  // If Frustum is successfully bulged, points on upper +zeta face
+  // should have the same radius:
+  const double radius = 4.0 * sqrt(3.0);
+  for (size_t i = 0; i < 4; i++) {
+    CHECK(magnitude(map(gsl::at(cornerpts, i))) == approx(4.0 * sqrt(3.0)));
```
Comment (nilsvu) [inline tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp]: Use `radius` variable here (same in other two functions below).

Comment (nilsvu) [inline tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp]: Can you also check the non-corner points?

**Review by nilsvu — CHANGES_REQUESTED** [review]

**Review by nilsvu — DISMISSED** [review]
LGTM, please squash 👍

**Review by nilsvu — DISMISSED** [review]

File: src/Domain/CoordinateMaps/Frustum.cpp [hunk 9]
```diff
@@ -201,7 +211,7 @@ std::optional<std::array<double, 3>> Frustum::inverse(
     logical_coords[2] = (-w_minus_ + w_plus_ * logical_coords[2]) /
                         (w_plus_ - w_minus_ * logical_coords[2]);
   }
-  if (sphericity_ > 0.0) {
+  if (sphericity_ > 0.0 or phi_ != 0.0) {
```
Comment (nikwit) [inline src/Domain/CoordinateMaps/Frustum.cpp:214]: When using this with the interpolation framework, I get an error because `gsl_multiroot` throws. This seems to happen because `block_logical_coordinates` iterates over all blocks and tries to find the block where the map is invertible by checking whether the inverse map returns a value or nullopt. So, I think we should catch this, something like:

```
    try {
      logical_coords = RootFinder::gsl_multiroot(
          rootfunction, logical_coords, absolute_tolerance, maximum_iterations,
          relative_tolerance, verbosity, max_absolute_tolerance, method,
          condition);

    } catch (const convergence_error& e) {
      return std::nullopt;
    }
```
Why your changes make this error show up I do not understand.

Comment (nilsvu) [reply src/Domain/CoordinateMaps/Frustum.cpp:214]: Is it possible to distinguish between cases where the point is outside the block and hence we _should_ return nullopt, and cases where the root find fails for other reasons and we should throw the error?

Comment (nikwit) [reply src/Domain/CoordinateMaps/Frustum.cpp:214]: I doubt there is a good way. One could check the convergence with iterations but that seems very unreliable.

On second thought, what about checking if the flat inverse is in the block before and returning nullopt if it is not?

On third thought, this is what we are already currently doing. So perhaps what has happened here is that the changes introduced in this PR mean that for certain configurations the initial guess / flat inverse will be inside the frustum even though it actually comes from a different block.

Comment (mar-celine) [reply src/Domain/CoordinateMaps/Frustum.cpp:214]: Hi Niko, I think I need a little help understanding what you encountered so that I know how to update the PR.
What were the parameters of the Frustum map where you saw this error? I am hoping that - if these parameters far from the ones used in the BinaryCompactObject Domain, that I could simply add a comment noting that for particular configurations, the multi-block inversion process is unstable.

Or, if this configuration is in the reasonable range of parameters, could you verify that it is the equiangular option that spoils the multi-block invertibility? If I have a set of parameters for which the frustum is multi-block invertible in the equidistant case and not in the equiangular case then maybe I can use that to narrow down what changes I should make. Thanks!

Comment (nikwit) [reply src/Domain/CoordinateMaps/Frustum.cpp:214]: Sure, I think I found a case:

Just taking the same frustum as in  `test_frustum_fail()` in `Test_Frustum:72` and giving it a value that is very far outside the frustum, it will throw a gsl error instead of returning `nullopt` or just inverting it. Can you reproduce this?

```
void test_frustum_fail_equiangular() {
  INFO("Frustum fail");
  const std::array<std::array<double, 2>, 4> face_vertices{
      {{{-2.0, -2.0}}, {{2.0, 2.0}}, {{-4.0, -4.0}}, {{4.0, 4.0}}}};
  const CoordinateMaps::Frustum map(
      face_vertices, 2.0, 5.0, OrientationMap<3>{});
  const std::array<double, 3> test_mapped_point4{{0., 50., 9.0}};
  map.inverse(test_mapped_point4); // works fine
  const CoordinateMaps::Frustum equiangular_map(
      face_vertices, 2.0, 5.0, OrientationMap<3>{}, true, 1.0, false, 1., 1.);
  equiangular_map.inverse(test_mapped_point4); // throws
}
```
Is it possible that the check that the denominators are positive (`Frustum.cpp:195`) may no longer be valid in the equiangular case?


Sorry for the long wait!

Comment (mar-celine) [reply src/Domain/CoordinateMaps/Frustum.cpp:214]: Hi Niko, thanks! I was able to reproduce the error - I think that the map inverse is failing not because of changes in this PR, because the test can be made to pass by reducing the sphericity to a value below `0.525`, independent of what the value for `phi` specified is.

I think the changes that need to be made are to add conditions so that `nullopt` is returned (and so a root find is not attempted) for a larger number of cases, since you found a case where a root find is attempted when it shouldn't be.

Since the map is nonlinear/ not analytically invertible at this point, I don't know if there's a clean way to do this that will catch all cases. I can add a short comment saying that it invertibility is not guaranteed for all Frustum configurations. I can also add a check that if the point to be inverted is outside the radius of the sphere containing the bulged frustum, to return nullopt

I'm not super familiar with how `block_logical_coordinates` uses the inverse map, will these fixes address the issue you ran into? Thanks for looking into this!

Comment (nikwit) [reply src/Domain/CoordinateMaps/Frustum.cpp:214]: Hey, 

I think it might be related to the equiangular map because for the `BinaryCompactObject` configuration that I am using this error popped up once I changed it to equiangular even though I was always using sphericity of 1 .

The problem is that `block_logical_coordinates` essentially just tries to invert the point using the inverse map of each block until it works so the inverse must either invert or return `nullopt` but it must not throw.

The simplest thing might be to just wrap the root finding in a `try` block and return nullopt if it fails. However, if you can come up with a way that checks whether the requested point is outside the block and returns `nullopt` in that case that is a lot cleaner of course. If you add this fix I am happy to try it on my `BinaryCompactObject` config.

**Review by nikwit — DISMISSED** [review]

File: src/Domain/CoordinateMaps/Frustum.cpp [hunk 10]
```diff
@@ -241,12 +259,16 @@ std::optional<std::array<double, 3>> Frustum::inverse(
...
-        relative_tolerance, verbosity, max_absolute_tolerance, method,
-        condition);
+    try {
+      logical_coords = RootFinder::gsl_multiroot(
+          rootfunction, logical_coords, absolute_tolerance, maximum_iterations,
+          relative_tolerance, verbosity, max_absolute_tolerance, method,
+          condition);
+
+    } catch (const convergence_error& e) {
+      return std::nullopt;
```
Comment (nilsvu) [inline src/Domain/CoordinateMaps/Frustum.cpp:269]: Can you add a tests that gives us confidence that the inversion works correctly? Perhaps create an equiangular frustum and probe a bunch of points inside, outside, and right on the boundary. I'm thinking that inverse(map(xi)) should always work for points inside and on the boundary, and inverse(x) should always return nullopt for points outside. 

Are there any points that lie inside the frustum if it is equidistant, but lie slightly outside if it is equiangular? I thought this was the cause for Niko's issue, but now I am unsure because the frustum doesn't actually change shape right? So I still don't quite understand the failure mode that Niko encountered.

Comment (mar-celine) [reply src/Domain/CoordinateMaps/Frustum.cpp:269]: Hi Nils, this should be covered ny the test suite for maps on the unit cube (At least, the testing of multiple points inside and on the boundaries of the Frustum). The test_frustum_fail tests that points on the outside of the Frustum, but do we want to return nullopt for every and all points outside [-1,1]^3? Because sometimes the point is invertible even when the point is outside, and the try catch block should cover the cases where it isn't. Is there a case you have in mind that isn't covered here?

As for equiangular vs equidistant, the equiangular map maps [-1,1] to itself, so points inside [-1,1]^3 should be inside in both cases, up to roundoff error.

**Comment by mar-celine** [comment]
@nikwit, @nilsvu has told me that this PR can be finished without him and I think I have address all the changes that have been requested up to this point, so hopefully we can approve this soon! :+1:

**Comment by nikwit** [comment]
I have already approved it, so do we need a different core dev to take a look at it?

**Comment by nilsdeppe** [comment]
I can take over as core dev :)

File: src/Domain/CoordinateMaps/Frustum.cpp (hunk 10 above)
Comment (nilsdeppe) [inline src/Domain/CoordinateMaps/Frustum.cpp:269]: Do you understand why code cov claims this isn't covered? Did you add a `std::cout` here to verify the failing example actually fails this check and not another? If it's a codecov issue that's fine, just want to make sure.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

**Review by nilsdeppe — APPROVED** [review]
